### ISSUE TITLE
Cursor offset + calibration issue in Beast Busters (bbusters)

### ISSUE THREAD

**Issue by StormedBubbles** [libretro/mame2003-plus-libretro#1222]
Thanks again, everyone, for all of your help lately!

I notice two issues with bbusters on a Pi 4 running RetroPie:

1. There is a cursor-tracking issue for all 3 players, whether using a gun or mouse. The shots go all over the place. The game does have a calibration menu for the guns, but...

2. The calibration menu doesn't seem to work. The menu is supposed to have 3 targets, but only 2 are visible. The procedure appears to be: calibrate each player individually by shooting at the 3 targets, and only after all 3 players have completed the calibration will the settings save. However, the ROM crashes when I try to turn off the service switch and restart the game, and any changes apparently fail to save.

**Comment by mahoneyt944** [comment]
I noticed this game yesterday, the internal gun reads need updated to track properly. Should be an easy fix, it's on my list of games to be fixed up.

**Comment by mahoneyt944** [comment]
Looks like a more recent fix up. https://github.com/mamedev/mame/commit/87d548d20590944336a0843c2741795bda04d6cc

Can't get the analog to play nicely but I may be able to patch this in another way.

**Comment by mahoneyt944** [comment]
I'll test if you have something. Can seem to get the masking right.

**Comment by arcadez2003** [comment]
No worries i'll send it later as i have to rejig some of the code for this core

**Comment by arcadez2003** [comment]
@mahoneyt944 Here ya go......
[bbusters.zip](https://github.com/libretro/mame2003-plus-libretro/files/7650693/bbusters.zip)

I cant guarantee there might not be typo or two in the mem maps but i checked everything real close
before sending, also need to run something by you.....

In the original driver if you look at the port maps you'll notice MEMORY_END where it should be PORT_END 
i dunno if that was an oversight by the dev that wrote the original driver or it was done for a reason.??
anyway since it should be PORT_END that's what i've used for the new maps but if it causes any issues
it's easy to revert.

Regards.

**Comment by mahoneyt944** [comment]
@arcadez2003 works well for me, great work.

@StormedBubbles you will have to delete bbusters.nv and bbusters.cfg, then update the core and rerun the rom.

These will be in the mame2003-plus save directory:

.../mame2003-plus/nvram/bbusters.nv
.../mame2003-plus/cfg/bbusters.cfg

**Comment by StormedBubbles** [comment]
bbusters seems good now! Thanks!

By default, the cursor lines up with my gun sight / mouse cursor. The third target in the calibration screen is restored. I think I may not be doing the calibration correctly because it keeps freezing on me after I complete it, but that doesn't really matter now that the aim is good by default.

**Comment by arcadez2003** [comment]
@mahoneyt944 check this commit if you dont mind.........

https://github.com/mamedev/mame/commit/67a5e53e17986679bc73d089e8c6c9993ae2b278#diff-c3af4d01d5f9d3de8f4c5127e296b11f7d5230f5efe75c0e1b2b714a89f3645d

Notice the extra ram reads for the service mode they were added after the mem maps were merged
AM_RAM can be both read and or write maybe we need to add a write for those to our ole map
so that the game does not lockup when exiting the calibration screen.?? as i noticed i only added
the reads which might be a mistake on my part back in the day.

**Comment by mahoneyt944** [comment]
Surely worth a try, though I think part of the issue is that the nvram loads a default calibration that's more compatible with our mame range than what actual hardware uses. I'm thinking that by using the calibration menu it might be detecting that our input is invalid and writing bad data. The note in your commit mentions this as well as the commit I linked above when the eeprom was removed and ranges were altered to match actual hardware.
https://github.com/libretro/mame2003-plus-libretro/blob/286c897505ac47af066c827da7d0e7c5b5404b74/src/drivers/bbusters.c#L167-L169
